File: bundle.go
package uipack

import (
	"bufio"
	"errors"
)

type Bundle struct {
	Version Version
	Variant Variant
	Values  []interface{}
}

func (variant *Bundle) Decode(r *bufio.Reader, metadata *BundleMetadata) error {
	protocol, err := readUint16(r)
	if err != nil {
		return err
	}
	if protocol != PROTOCOL_VERSION {
		return errors.New("unsupported protocol version")
	}
	err = variant.Version.Decode(r)
	if err != nil {
		return err
	}
	v, err := readUint64(r)
	if err != nil {
		return err
	}
	variant.Variant = Variant(v)
	values := make([]interface{}, len(metadata.Variables))

	for i, variable := range metadata.Variables {
		values[i], err = metadata.DecodeArgument(variable.Type, r)
		if err != nil {
			return err
		}
	}

	variant.Values = values
	return nil
}

func (variant *Bundle) Encode(w *bufio.Writer, metadata *BundleMetadata) error {
	err := writeUint16(w, PROTOCOL_VERSION)
	if err != nil {
		return err
	}
	err = variant.Version.Encode((*bufio.Writer)(w))
	if err != nil {
		return err
	}
	err = writeUint64(w, uint64(variant.Variant))
	if err != nil {
		return err
	}
	for _, variable := range variant.Values {
		err = metadata.EncodeArgument(w, variable)
		if err != nil {
			return err
		}
	}
	return nil
}

func (metadata *BundleMetadata) DecodeArgument(t ValueType, r *bufio.Reader) (interface{}, error) {
	switch MainValueType(t.Type) {
	case IntegerType:
		return readUint64(r)
	case BooleanType:
		return readBool(r)
	case FloatType:
		return readFloat64(r)
	case StringType:
		return readString(r)
	case ColorType:
		result := Color{}
		err := result.Decode(r)
		return result, err
	case TextStyleType:
		result := TextStyle{}
		err := result.Decode(r)
		return result, err
	case LinearGradientType:
		result := LinearGradient{}
		err := result.Decode(r)
		return result, err
	case RadialGradientType:
		result := RadialGradient{}
		err := result.Decode(r)
		return result, err
	case LabelType:
		result := Label{}
		err := result.Decode(r)
		return result, err
	case OffsetType:
		result := Offset{}
		err := result.Decode(r)
		return result, err
	case RadiusType:
		result := Radius{}
		err := result.Decode(r)
		return result, err
	case BorderRadiusType:
		result := BorderRadius{}
		err := result.Decode(r)
		return result, err
	case CustomType:
		result := Instance{}
		err := result.Decode(r, metadata)
		return result, err
	default:
		return nil, errors.New("unknown type")
	}
}

func (metadata *BundleMetadata) EncodeArgument(w *bufio.Writer, data interface{}) error {
	switch v := data.(type) {
	case Color:
		return v.Encode(w)
	case TextStyle:
		return v.Encode(w)
	case LinearGradient:
		return v.Encode(w)
	case RadialGradient:
		return v.Encode(w)
	case Label:
		return v.Encode(w)
	case Offset:
		return v.Encode(w)
	case Instance:
		return v.Encode(w, metadata)
	case Radius:
		return v.Encode(w)
	case BorderRadius:
		return v.Encode(w)
	case uint64:
		return writeUint64(w, v)
	case bool:
		return writeBool(w, v)
	case float64:
		return writeFloat64(w, v)
	case string:
		return writeString(w, v)
	}

	return errors.New("unknown type")
}

File: packages/uipack/lib/src/models/bundle.dart
import '../encoding/binary_encoding.dart';
import 'version.dart';
import 'variant.dart';
import 'metadata.dart';
import 'color.dart';
import 'text_style.dart';

class Bundle {
  Version version;
  Variant variant;
  List<dynamic> values;

  Bundle({
    required this.version,
    required this.variant,
    required this.values,
  });

  void decode(ByteDataReader reader, BundleMetadata metadata) {
    final protocol = reader.readUint16();
    if (protocol != protocolVersion) {
      throw Exception('Unsupported protocol version: $protocol');
    }

    version = Version(major: 0, minor: 0);
    version.decode(reader);
    
    final variantValue = reader.readUint64();
    variant = Variant(variantValue);
    
    values = List.generate(metadata.variables.length, (i) {
      final variable = metadata.variables[i];
      return _decodeArgument(variable.type, reader, metadata);
    });
  }

  void encode(ByteDataWriter writer, BundleMetadata metadata) {
    writer.writeUint16(protocolVersion);
    version.encode(writer);
    writer.writeUint64(variant.value);
    
    for (final value in values) {
      _encodeArgument(writer, value, metadata);
    }
  }

  dynamic _decodeArgument(ValueType type, ByteDataReader reader, BundleMetadata metadata) {
    switch (type.type) {
      case MainValueType.integer:
        return reader.readUint64();
      case MainValueType.boolean:
        return reader.readBool();
      case MainValueType.float:
        return reader.readFloat64();
      case MainValueType.string:
        return reader.readString();
      case MainValueType.color:
        final color = Color(
          red: 0.0,
          green: 0.0,
          blue: 0.0,
          alpha: 1.0,
          colorSpace: ColorSpace.srgb,
        );
        color.decode(reader);
        return color;
      case MainValueType.textStyle:
        final textStyle = TextStyle(
          fontFamily: '',
          fontSize: 0.0,
          fontWeight: 400,
          letterSpacing: 0.0,
          wordSpacing: 0.0,
          lineHeight: 1.0,
          fontVariations: [],
        );
        textStyle.decode(reader);
        return textStyle;
      default:
        throw Exception('Unsupported type: ${type.type}');
    }
  }

  void _encodeArgument(ByteDataWriter writer, dynamic data, BundleMetadata metadata) {
    switch (data.runtimeType) {
      case Color:
        (data as Color).encode(writer);
        break;
      case TextStyle:
        (data as TextStyle).encode(writer);
        break;
      case int:
        writer.writeUint64(data as int);
        break;
      case bool:
        writer.writeBool(data as bool);
        break;
      case double:
        writer.writeFloat64(data as double);
        break;
      case String:
        writer.writeString(data as String);
        break;
      default:
        throw Exception('Unsupported type: ${data.runtimeType}');
    }
  }
}